From 431fb3222d52c82dd35e37b35e887cd363c2b825 Mon Sep 17 00:00:00 2001
From: Stelios Tsampas <loathingkernel@gmail.com>
Date: Wed, 17 Jun 2026 01:12:39 +0300
Subject: [PATCH 4/6] upscalers: use a single version tracking file

---
 upscalers.py | 105 +++++++++++++++++++++++++++++----------------------
 1 file changed, 60 insertions(+), 45 deletions(-)

diff --git a/upscalers.py b/upscalers.py
index c1ab5fd..f1abd60 100644
--- a/upscalers.py
+++ b/upscalers.py
@@ -79,12 +79,13 @@ def __dll_download_exists(url: str) -> bool:
     return False
 
 
-__dlss_version_file = 'dlss_version'
-__xess_version_file = 'xess_version'
-__fsr3_version_file = 'fsr3_version'
-__fsr4_version_file = 'fsr4_version'
-__opti_version_file = 'opti_version'
-__fdfx_version_file = 'fdfx_version'
+__dlss_section = 'dlss_files'
+__xess_section = 'xess_files'
+__fsr3_section = 'fsr3_files'
+__fsr4_section = 'fsr4_files'
+__opti_section = 'opti_files'
+__fdfx_section = 'fdfx_files'
+__version_file = 'upscaler_files'
 
 
 def __get_opti_files(version: str = 'default') -> tuple[str, dict]:
@@ -205,13 +206,13 @@ def __get_fsr4_dlls(version: str = 'default') -> dict:
 
 def __get_upscaler_items(name: str, version: str) -> tuple[dict, str]:
     upscalers = {
-        'dlss': (__get_dlss_dlls, __dlss_version_file),
-        'xess': (__get_xess_dlls, __xess_version_file),
-        'fsr3': (__get_fsr3_dlls, __fsr3_version_file),
-        'fsr4': (__get_fsr4_dlls, __fsr4_version_file),
-        'fdfx': (__get_fdfx_dlls, __fdfx_version_file),
+        'dlss': (__get_dlss_dlls, __dlss_section),
+        'xess': (__get_xess_dlls, __xess_section),
+        'fsr3': (__get_fsr3_dlls, __fsr3_section),
+        'fsr4': (__get_fsr4_dlls, __fsr4_section),
+        'fdfx': (__get_fdfx_dlls, __fdfx_section),
     }
-    get_items, version_file = upscalers[name]
+    get_items, section = upscalers[name]
     try:
         items = get_items(version)
     except Exception as e:
@@ -219,14 +220,16 @@ def __get_upscaler_items(name: str, version: str) -> tuple[dict, str]:
         log.crit(e)
         raise e
 
-    return items, version_file
+    return items, section
 
 
-def __get_local_versions(version_file: str) -> dict:
+def __get_local_versions(compat_dir: str, section: str) -> dict:
+    version_file = os.path.join(compat_dir, __version_file)
     try:
         with open(version_file, encoding='utf-8') as version_fd:
             data = version_fd.read()
         local_versions = json.loads(data)
+        local_versions = local_versions[section]
     except Exception as e:
         log.warn(f'Error while reading version file "{version_file}"')
         raise e
@@ -234,6 +237,22 @@ def __get_local_versions(version_file: str) -> dict:
     return local_versions
 
 
+def __set_local_versions(compat_dir: str, section: str, checksums: dict) -> None:
+    version_file = os.path.join(compat_dir, __version_file)
+    try:
+        with open(version_file, encoding='utf-8') as version_fd:
+            data = version_fd.read()
+        local_versions = json.loads(data)
+    except Exception as e:
+        log.warn(f'Error while reading version file "{version_file}"')
+        local_versions = {}
+
+    local_versions[section] = checksums
+
+    with open(version_file, 'w', encoding='utf-8') as version_fd:
+        version_fd.write(json.dumps(local_versions))
+
+
 def __check_upscaler_file(
     prefix_dir: str, dst: str, item: dict, version: dict, ignore_version: bool
 ) -> bool:
@@ -278,10 +297,10 @@ def __check_upscaler_file(
 
 
 def __check_upscaler_files(
-    prefix_dir: str, items: dict, version_file: str, ignore_version: bool
+    compat_dir: str, prefix_dir: str, items: dict, section: str, ignore_version: bool
 ) -> bool:
     try:
-        local_versions = __get_local_versions(version_file)
+        local_versions = __get_local_versions(compat_dir, section)
         # test if new files and their attributes exist in the tracking file
         for dst in items.keys():
             _ = local_versions[dst].get('md5_hash')
@@ -312,24 +331,25 @@ def check_upscaler(
     ignore_version: ignore version mismatch but still check if the dlls are present
     """
     try:
-        items, version_file = __get_upscaler_items(name, version)
+        items, section = __get_upscaler_items(name, version)
     except Exception:
         return False
 
     return __check_upscaler_files(
+        compat_dir,
         prefix_dir,
         items,
-        os.path.join(compat_dir, version_file),
+        section,
         ignore_version,
     )
 
 
 def check_optiscaler(
-        compat_dir: str,
-        prefix_dir: str,
-        version: str = 'default',
-        *,
-        ignore_version: bool = False,
+    compat_dir: str,
+    prefix_dir: str,
+    version: str = 'default',
+    *,
+    ignore_version: bool = False,
 ) -> bool:
     """Check for optiscaler files and its version
 
@@ -343,9 +363,8 @@ def check_optiscaler(
         log.crit(e)
         return False
 
-    version_file = os.path.join(compat_dir, __opti_version_file)
     try:
-        local_versions = __get_local_versions(version_file)
+        local_versions = __get_local_versions(compat_dir, __opti_section)
         # test if new files and their attributes exist in the tracking file
         for dst in [os.path.join(path, f) for f in item['md5_hash']]:
             _ = local_versions[dst].get('md5_hash')
@@ -387,10 +406,11 @@ def check_optiscaler(
 
 
 def __download_upscaler_files(
+    compat_dir: str,
     prefix_dir: str,
     items: dict,
     dlfunc: Callable[[dict, Path, Path], None],
-    version_file: str,
+    section: str,
 ) -> bool:
     """Download and install the required dlls.
 
@@ -423,8 +443,7 @@ def __download_upscaler_files(
             'version': items[dst]['version'],
             'md5_hash': items[dst]['md5_hash'],
         }
-    with open(version_file, 'w', encoding='utf-8') as version_fd:
-        version_fd.write(json.dumps(version))
+    __set_local_versions(compat_dir, section, version)
     return True
 
 
@@ -513,20 +532,21 @@ def download_upscaler(
     log.info(f'Failed to validate "{name.upper()}" files.')
 
     upscalers = {
-        'dlss': (__get_dlss_dlls, __download_extract_zip, __dlss_version_file),
-        'xess': (__get_xess_dlls, __download_extract_zip, __xess_version_file),
-        'fsr3': (__get_fsr3_dlls, __download_extract_zip, __fsr3_version_file),
-        'fsr4': (__get_fsr4_dlls, __download_fsr4, __fsr4_version_file),
-        'fdfx': (__get_fdfx_dlls, __download_extract_zip, __fdfx_version_file),
+        'dlss': (__get_dlss_dlls, __download_extract_zip, __dlss_section),
+        'xess': (__get_xess_dlls, __download_extract_zip, __xess_section),
+        'fsr3': (__get_fsr3_dlls, __download_extract_zip, __fsr3_section),
+        'fsr4': (__get_fsr4_dlls, __download_fsr4, __fsr4_section),
+        'fdfx': (__get_fdfx_dlls, __download_extract_zip, __fdfx_section),
     }
-    get_items, download_func, version_file = upscalers[name]
+    get_items, download_func, section = upscalers[name]
     try:
         items = get_items(version)
         if not __download_upscaler_files(
+            compat_dir,
             prefix_dir,
             items,
             download_func,
-            os.path.join(compat_dir, version_file),
+            section,
         ):
             raise RuntimeError
     except Exception as e:
@@ -545,9 +565,8 @@ def download_optiscaler(
         return
     log.info('Failed to validate "optiscaler" files.')
 
-    version_file = os.path.join(compat_dir, __opti_version_file)
     try:
-        local_versions = __get_local_versions(version_file)
+        local_versions = __get_local_versions(compat_dir, __opti_section)
     except Exception:
         local_versions = {}
         pass
@@ -590,8 +609,7 @@ def download_optiscaler(
         log.crit(e)
         return
 
-    with open(version_file, 'w', encoding='utf-8') as version_fd:
-        version_fd.write(json.dumps(version))
+    __set_local_versions(compat_dir, __opti_section, version)
 
 
 def setup_upscaler(
@@ -623,21 +641,18 @@ def clean_upscaler(
     prefix_dir: str,
 ) -> bool:
     try:
-        _, version_file = __get_upscaler_items(name, 'default')
+        _, section = __get_upscaler_items(name, 'default')
     except Exception:
         return False
 
     log.info(f'Cleaning {name.upper()} files.')
 
-    version_file = os.path.join(compat_dir, version_file)
     try:
-        local_versions = __get_local_versions(version_file)
+        local_versions = __get_local_versions(compat_dir, section)
     except Exception:
         return False
 
-    if os.path.exists(version_file):
-        log.debug(f'Removing {version_file}')
-        os.unlink(version_file)
+    __set_local_versions(compat_dir, section, {})
 
     for path in local_versions:
         file = os.path.join(prefix_dir, path)
-- 
2.54.0

