From 397930db8be853132d6f5434adaafc91a3289920 Mon Sep 17 00:00:00 2001
From: Stelios Tsampas <loathingkernel@gmail.com>
Date: Thu, 14 May 2026 03:10:08 +0300
Subject: [PATCH 3/5] upscalers: use context managers with read()

---
 upscalers.py | 10 +++++++---
 1 file changed, 7 insertions(+), 3 deletions(-)

diff --git a/upscalers.py b/upscalers.py
index 6705649..629b776 100644
--- a/upscalers.py
+++ b/upscalers.py
@@ -418,7 +418,8 @@ def __download_file(url: str, dst: Path, *, checksum: Union[str, None] = None) -
         with dst.open('wb') as dst_fd:
             with urllib.request.urlopen(request, timeout=10) as url_fd:
                 dst_fd.write(url_fd.read())
-        dst_md5 = hashlib.md5(dst.open('rb').read()).hexdigest().lower()
+        with dst.open('rb') as dst_fd:
+            dst_md5 = hashlib.md5(dst_fd.read()).hexdigest().lower()
         dst_size = dst.stat().st_size if dst.exists() else 0
         # Size check is arbitrary, but nothing should be below 1K
         if (checksum and dst_md5 != checksum.lower()) or dst_size < 1024:
@@ -432,7 +433,8 @@ def __cached_download(item: dict, cached_file: Path) -> None:
     item_md5 = item.get('zip_md5_hash', '')
 
     if cached_file.exists():
-        cached_md5 = hashlib.md5(cached_file.open('rb').read()).hexdigest().lower()
+        with cached_file.open('rb') as cached_fd:
+            cached_md5 = hashlib.md5(cached_fd.read()).hexdigest().lower()
         if item_md5 and cached_md5 != item_md5.lower():
             log.crit(f'MD5 mismatch between manifest and cached "{cached_file.name}"')
             cached_file.unlink(missing_ok=True)
@@ -451,7 +453,9 @@ def __download_extract_zip(item: dict, cache: Path, dst: Path) -> None:
             zip_fd.extractall(dst.parent)
     if cached_file.suffix == '.xz':
         with dst.open('wb') as dst_fd:
-            dst_fd.write(lzma.decompress(cached_file.open('rb').read()))
+            # this also sets the target filename
+            with cached_file.open('rb') as cached_fd:
+                dst_fd.write(lzma.decompress(cached_fd.read()))
 
 
 def __download_fsr4(item: dict, cache: Path, dst: Path) -> None:
-- 
2.54.0

