From fa4e3dad8310e47d103ca6adebed57b68dbe0df5 Mon Sep 17 00:00:00 2001
From: Stelios Tsampas <loathingkernel@gmail.com>
Date: Sun, 15 Feb 2026 14:24:53 +0200
Subject: [PATCH 2/6] fix: add early stage fixes to run when the module is
 imported

---
 __init__.py | 13 ++++++++++++
 fix.py      | 57 ++++++++++++++++++++++++++++++++++++-----------------
 2 files changed, 52 insertions(+), 18 deletions(-)

diff --git a/__init__.py b/__init__.py
index dca1982..498cfeb 100644
--- a/__init__.py
+++ b/__init__.py
@@ -62,6 +62,19 @@ def check_iscriptevaluator() -> bool:
     return len(sys.argv) >= 3 and 'iscriptevaluator.exe' in sys.argv[2]
 
 
+def execute_early() -> None:
+    """Execute the early part of protonfixes"""
+    if check_iscriptevaluator():
+        log.debug('Skipping fix execution. We are running "iscriptevaluator.exe".')
+    elif not check_conditions():
+        log.warn('Skipping fix execution. We are probably running a unit test.')
+    else:
+        fix.early()
+
+
+execute_early()
+
+
 def setup(
     env: dict,
     bin_path_var: str,
diff --git a/fix.py b/fix.py
index 31d2673..9584880 100644
--- a/fix.py
+++ b/fix.py
@@ -116,7 +116,7 @@ def get_module_name(game_id: str, default: bool = False, local: bool = False) ->
     )
 
 
-def _run_fix_local(game_id: str, default: bool = False) -> bool:
+def _run_fix_local(game_id: str, stage: str, default: bool = False) -> bool:
     """Check if a local gamefix is available first and run it"""
     localpath = os.path.expanduser('~/.config/protonfixes/localfixes')
     module_name = game_id if not default else 'default'
@@ -130,10 +130,10 @@ def _run_fix_local(game_id: str, default: bool = False) -> bool:
         sys.path.append(os.path.expanduser('~/.config/protonfixes'))
 
     # Run fix
-    return _run_fix(game_id, default, True)
+    return _run_fix(game_id, stage, default, True)
 
 
-def _run_fix(game_id: str, default: bool = False, local: bool = False) -> bool:
+def _run_fix(game_id: str, stage: str, default: bool = False, local: bool = False) -> bool:
     """Private function, which actually executes gamefixes"""
     fix_type = 'protonfix' if not default else 'defaults'
     scope = 'global' if not local else 'local'
@@ -142,18 +142,25 @@ def _run_fix(game_id: str, default: bool = False, local: bool = False) -> bool:
         module_name = get_module_name(game_id, default, local)
         game_module = import_module(module_name)
 
-        log.info(f'Using {scope} {fix_type} for {get_game_name()} ({game_id})')
-        if hasattr(game_module, 'main_with_id'):
-            game_module.main_with_id(game_id)
-        else:
-            game_module.main()
+        log.info(f'Using {stage} stage {scope} {fix_type} for {get_game_name()} ({game_id})')
+        if stage == 'early':
+            if hasattr(game_module, 'early_with_id'):
+                game_module.early_with_id(game_id)
+            elif hasattr(game_module, 'early'):
+                game_module.early()
+
+        elif stage == 'main':
+            if hasattr(game_module, 'main_with_id'):
+                game_module.main_with_id(game_id)
+            else:
+                game_module.main()
     except ImportError:
-        log.info(f'No {scope} {fix_type} found for {get_game_name()} ({game_id})')
+        log.info(f'No {stage} stage {scope} {fix_type} found for {get_game_name()} ({game_id})')
         return False
     return True
 
 
-def run_fix(game_id: str) -> None:
+def run_fix(game_id: str, *, stage: str) -> None:
     """Loads a gamefix module by it's gameid
 
     local fixes prevent global fixes from being executed
@@ -165,16 +172,16 @@ def run_fix(game_id: str) -> None:
         run_checks()
 
     # execute default.py (local)
-    if not _run_fix_local(game_id, True) and config.main.enable_global_fixes:
-        _run_fix(game_id, True)  # global
+    if not _run_fix_local(game_id, stage, True) and config.main.enable_global_fixes:
+        _run_fix(game_id, stage, True)  # global
 
     # execute <game_id>.py (local)
-    if not _run_fix_local(game_id, False) and config.main.enable_global_fixes:
-        _run_fix(game_id, False)  # global
+    if not _run_fix_local(game_id, stage, False) and config.main.enable_global_fixes:
+        _run_fix(game_id, stage, False)  # global
 
 
-def main() -> None:
-    """Runs the gamefix"""
+def _entry_check() -> bool:
+    """Checks if the gamefix should run"""
     check_args = [
         'iscriptevaluator.exe' in sys.argv[2],
         'getcompatpath' in sys.argv[1],
@@ -184,10 +191,24 @@ def main() -> None:
     if any(check_args):
         log.debug(str(sys.argv))
         log.debug('Not running protonfixes for setup runs')
-        return
+        return False
 
     version = ProtonVersion.from_version_file()
     log.info(
         f'Running protonfixes on "{version.version_name}", build at {version.build_date}.'
     )
-    run_fix(get_game_id())
+    return True
+
+
+def early() -> None:
+    """Runs the early part of a gamefix"""
+    if not _entry_check():
+        return
+    run_fix(get_game_id(), stage='early')
+
+
+def main() -> None:
+    """Runs the main part of a gamefix"""
+    if not _entry_check():
+        return
+    run_fix(get_game_id(), stage='main')
-- 
2.53.0

